cxx/isce3/core/Utilities.h

```diff
@@ -339,17 +339,25 @@ namespace isce3 { namespace core {
         int left = 0;
         int right = array.size() - 1;
         int index;
-        while (left <= right) {
+
+        long long count = 0;
+        long long MAX_NUM_ITERATIONS = 1e11;
+        while (left <= right and count < MAX_NUM_ITERATIONS) {
             const int middle = static_cast<int>(std::round(0.5 * (left + right)));
+            const auto middle_value = array[middle];
+            if (std::isnan(middle_value)) {
+                throw std::invalid_argument("input array may not contain NaN values");
+            }
             if (left == (right - 1)) {
                 index = left;
                 return index;
             }
-            if (array[middle] <= value) {
+            if (middle_value <= value) {
                 left = middle;
-            } else if (array[middle] > value) {
+            } else if (middle_value > value) {
```

Comment:
Technically we don't need this `if` condition, I think.

If `middle_value` is not NaN and `middle_value` is not `<= value` (checked above on line 355), then `middle_value > value` must be true (for `double`, at least).

```suggestion
            } else {
```

Reply:
Sounds good. I wonder if that would make any difference in the compiled code. Anyways, I updated it in the commit https://github.com/isce-framework/isce3/pull/53/commits/feef376d97df4485b7a3bb9b3cb81b8b4497a4d4. Thanks!